cl/api/utils.py

```diff
@@ -540,3 +545,123 @@ def get_avg_ms_for_endpoint(endpoint: str, d: datetime) -> float:
     results = pipe.execute()
 
     return results[0] / results[1]
+
+
+def get_next_webhook_retry_date(retry_counter: int) -> datetime:
+    """Returns the next retry datetime to schedule a webhook retry based on its
+    current retry counter.
+
+    The next retry date is computed based to meet an exponential backoff with a
+    starting multiplier of three.
+
+      Retry counter | New Delay
+      -------------+------------
+            1      |     0:03
+            2      |     0:09
+            3      |     0:27
+            4      |     1:21
+            5      |     4:03
+            6      |     12:09
+            7      |     36:27
+
+
+    The total elapsed time might vary for each webhook event depending on when
+    the retry is executed since the retry method is going to be executed every
+    minute. On average the total elapsed time in the 7th retry would be 54
+    hours and 39 minutes.
+
+    :param retry_counter: The current retry_counter used to compute the next
+    retry date.
+    :return datatime: The next retry datetime.
+    """
+
+    INITIAL_TIME = 3  # minutes
+    # Update new_next_retry_date exponentially
+    new_next_retry_date = now() + timedelta(
+        minutes=pow(INITIAL_TIME, retry_counter)
+    )
+    return new_next_retry_date
+
+
+def update_webhook_event_after_request(
+    webhook_event: WebhookEvent, response: Response = None, error: str = ""
+) -> None:
+    """Update the webhook event after sending the POST request. If the webhook
+    event fails, increase the retry counter, next retry date and increase its
+    parent webhook failure count. If the webhook event reaches the max retry
+    counter marks it as Failed. If there is no error marks it as Successful.
+
+    :param webhook_event: The WebhookEvent to update.
+    :param response: Optional in case we receive a requests Response object, to
+    update the WebhookEvent accordingly.
+    :param error: Optional, if we don't receive a request Response we would
+    receive an error to log it.
+    :return: None
+    """
+
+    WEBHOOK_MAX_RETRY_COUNTER = 7
+
+    failed_request = False
+    if response is not None:
+        webhook_event.status_code = response.status_code
+        webhook_event.response = response.text
+        # If the response status code is not between 200 and 400, it's
+        # considered a failed attempt and will be enqueued for retry.
+        if not response.ok:
```

Comment:
I forget. Did we decide redirects were allowed? I forget what the norm is for this, but I think if we support them, we need to set `follow=True` on our post. If we don't support them, then we need to tweak `response.ok` to be more strict. 

I *think* it's probably better not to support them, so that people don't get used to having them and we don't have to spend time following them.

Reply:
Yeah, you're right we're not going to support redirects as discussed in #2273. So I will change it to avoid redirects.

Reply:
Done, we won't allow redirects and if we receive a 300 status the webhook event will be marked as Failed. Only `2xx` status codes are considered a successful delivery.